File: code/91/Java/dp2.java
class Solution {
    public int numDecodings(String s) {
        if (s==null || s.length()==0) return 0;
        char[] ca = s.toCharArray();
        int dp[] = new int[s.length()];
        if (ca[0]!='0') dp[0]=1;
        for (int i=1; i<s.length(); i++) {
            if (ca[i]!='0') dp[i] += dp[i-1];
            if (ca[i-1]=='1' || (ca[i-1]=='2' && ca[i]<='6')) dp[i] += (i-2>=0? dp[i-2]: 1);
        }
        return dp[s.length()-1];
    }
}

File: code/91/Python/dp2.py
class Solution:
    def numDecodings(self, s):
        if not s: return 0
        dp = [0]*(len(s)+1)
        dp[-1] = 1
        if s[0]!='0': dp[0]=1
        for i in range(1, len(s)):
            if s[i]!='0': dp[i] += dp[i-1]
            if s[i-1]=='1' or (s[i-1]=='2' and s[i]<='6'): dp[i] += dp[i-2]
        return dp[len(s)-1]